### PR TITLE
raftstore: enable v1 to receive snapshot from v2

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #14579

What's Changed:

```commit-message
enable v1 to receive snapshot from v2
```

This PR enables raft peer with config of raftstore-v1 to receive snapshot from peers of raftstore-v2.
Now, the raftstore cannot use the snapshot received from v2 properly, which will be done in the following PR(s).

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/server/snap.rs [hunk 1]
```diff
@@ -364,6 +364,9 @@ pub struct Runner<R: RaftExtension> {
     cfg: Config,
     sending_count: Arc<AtomicUsize>,
     recving_count: Arc<AtomicUsize>,
+
+    // only used when the snapshot is setn from raftstore-v2
```
Comment (bufferflies) [inline src/server/snap.rs:368]:
```suggestion
    // only used when the snapshot is sent from raftstore-v2
```

File: src/server/snap.rs [hunk 2]
```diff
@@ -364,6 +364,9 @@ pub struct Runner<R: RaftExtension> {
...
+    limiter: Limiter,
```
Comment (bufferflies) [inline src/server/snap.rs:369]: why not use the `snap_mgr`'s limit?

Comment (SpadeA-Tang) [reply src/server/snap.rs:369]: Good catch. I miss it.

File: components/test_raftstore-v2/src/util.rs [hunk 3]
```diff
@@ -233,3 +241,32 @@ pub fn async_read_on_peer<T: Simulator<EK>, EK: KvEngine>(
     request.mut_header().set_replica_read(replica_read);
     cluster.sim.wl().async_read(request)
 }
+
+pub fn call_conf_change<T>(
```
Comment (bufferflies) [inline components/test_raftstore-v2/src/util.rs:245]: where it is used?

File: components/raftstore/src/store/snap.rs [hunk 4]
```diff
@@ -1387,13 +1387,17 @@ struct SnapManagerCore {
 pub struct SnapManager {
     core: SnapManagerCore,
     max_total_size: Arc<AtomicU64>,
+
+    // used to receive snapshot from v2
+    core_v2: TabletSnapManager,
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore/src/store/snap.rs]: Good to me.

File: components/test_raftstore-v2/src/util.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/test_raftstore-v2/src/util.rs]: It is not needed. I have removed it.

File: src/server/snap.rs [hunk 5]
```diff
@@ -451,12 +459,35 @@ impl<R: RaftExtension + 'static> Runnable for Runner<R> {
                 };
                 self.pool.spawn(task);
             }
-            Task::RecvTablet { sink, .. } => {
-                let status = RpcStatus::with_message(
-                    RpcStatusCode::UNIMPLEMENTED,
-                    "tablet snap is not supported".to_string(),
-                );
-                self.pool.spawn(sink.fail(status).map(|_| ()));
+            Task::RecvTablet { stream, sink } => {
```
Comment (overvenus) [inline src/server/snap.rs:462]: It should only receive tablet snapshot when it's tiflash proxy.

File: tests/integrations/raftstore/test_snap.rs [hunk 6]
```diff
@@ -733,3 +738,101 @@ fn test_snapshot_clean_up_logs_with_log_gc() {
...
+                .await
+        });
+
+        // The snapshot has been received by cluster v1, so check it's completeness
+        let snap_mgr = cluster_v1.get_snap_mgr(1);
+        let path = snap_mgr.tablet_snap_manager().final_recv_path(&snap_key);
+        let rocksdb = engine_rocks::util::new_engine_opt(
+            path.as_path().to_str().unwrap(),
+            RocksDbOptions::default(),
+            vec![("default", RocksCfOptions::default())],
```
Comment (overvenus) [inline tests/integrations/raftstore/test_snap.rs:818]: What about `write` and `lock`?

Comment (SpadeA-Tang) [reply tests/integrations/raftstore/test_snap.rs:818]: no data was written in write and lock, so it's unnecessary to open them here.

File: src/server/snap.rs (hunk 5 above)
Comment (overvenus) [reply src/server/snap.rs:462]: You can check store label, `engine = tiflash`.

File: tests/integrations/raftstore/test_snap.rs (hunk 6 above)
Comment (overvenus) [reply tests/integrations/raftstore/test_snap.rs:824]: It's best to check if both of them exist for completeness.

Comment (SpadeA-Tang) [reply tests/integrations/raftstore/test_snap.rs:824]: Fine

File: src/server/snap.rs [hunk 7]
```diff
@@ -451,12 +466,44 @@ impl<R: RaftExtension + 'static> Runnable for Runner<R> {
...
+                    let status = RpcStatus::with_message(
+                        RpcStatusCode::UNIMPLEMENTED,
+                        "tablet snap is not supported".to_string(),
+                    );
+                    self.pool.spawn(sink.fail(status).map(|_| ()));
+                    return;
+                }
+
+                if let Some(status) = self.receiving_busy() {
+                    self.pool.spawn(sink.fail(status));
```
Comment (bufferflies) [inline src/server/snap.rs:480]: why is it difference between the code line 475?

Comment (SpadeA-Tang) [reply src/server/snap.rs:480]: The original code is different. I don't know why.

File: components/raftstore/src/store/snap.rs [hunk 8]
```diff
@@ -1896,9 +1912,12 @@ impl SnapManagerBuilder {
         } else {
             u64::MAX
         };
+        let path = path.into();
+        let mut path_v2 = path.clone();
+        path_v2.push_str("_v2");
```
Comment (overvenus) [inline components/raftstore/src/store/snap.rs:1917]: Does it cause compatibility issues? E.g., upgrade from an old tikv, can it read or clean up old snapshots.

Comment (SpadeA-Tang) [reply components/raftstore/src/store/snap.rs:1917]: this is the path for v2's snap mgr, v1's snap mgr will not be changed.

**Review by overvenus — APPROVED** [review]

**Review by bufferflies — APPROVED** [review]

File: tests/integrations/raftstore/test_snap.rs [hunk 9]
```diff
@@ -733,3 +738,110 @@ fn test_snapshot_clean_up_logs_with_log_gc() {
...
+
+        for i in 0..key_num {
+            let k = format!("zk{:04}", i);
+            assert!(
+                rocksdb
+                    .get_value_cf("default", k.as_bytes())
+                    .unwrap()
+                    .is_some()
+            );
+        }
```
Comment (CalvinNeo) [inline tests/integrations/raftstore/test_snap.rs:839]: Should we explicitly call shutdown here?

Comment (SpadeA-Tang) [reply tests/integrations/raftstore/test_snap.rs:839]: why?

**Pull request merged**